### PR TITLE
Remove offline sessions when deleting a realm

### LINKED ISSUE #13110: Remove offline sessions when deleting a realm
When a realm is removed, offline sessions remain in a storage. They should be removed.

### PR DESCRIPTION
**Pull request by martin-kanis**
Fixes #13110

### CODE REVIEW

**Comment by martin-kanis** [comment]
We should check invalidation handling for all other areas within https://github.com/keycloak/keycloak/issues/11842 issue.

File: model/map/src/main/java/org/keycloak/models/map/userSession/MapUserSessionProviderFactory.java [hunk 1]
```diff
@@ -64,6 +51,11 @@ public String getHelpText() {
     public void invalidate(KeycloakSession session, InvalidableObjectType type, Object... params) {
         if (type == USER_BEFORE_REMOVE) {
             create(session).removeUserSessions((RealmModel) params[0], (UserModel) params[1]);
+        } else if (type == REALM_BEFORE_REMOVE) {
+            MapUserSessionProvider mapUserSessionProvider = create(session);
+
+            mapUserSessionProvider.removeUserSessions((RealmModel) params[0]);
+            mapUserSessionProvider.removeOfflineUserSessions((RealmModel) params[0]);
```
Comment (mhajas) [inline model/map/src/main/java/org/keycloak/models/map/userSession/MapUserSessionProviderFactory.java:58]: Can this be done using only one storage call?

**Review by mhajas — CHANGES_REQUESTED** [review]
Thanks for the PR @martin-kanis. I added one comment in the code. Also, can we create some tests for this to avoid similar issues later? Maybe using `HotRodConnectionProvider` and looking directly into the cache? Perhaps it could be enough to test using one implementation.

File: model/map/src/main/java/org/keycloak/models/map/userSession/MapUserSessionProviderFactory.java (hunk 1 above)
Comment (martin-kanis) [reply model/map/src/main/java/org/keycloak/models/map/userSession/MapUserSessionProviderFactory.java]: Yes, I changed the method to remove both online and offline sessions.

File: testsuite/model/src/test/java/org/keycloak/testsuite/model/session/OfflineSessionPersistenceTest.java [hunk 2]
```diff
@@ -122,6 +127,37 @@ public void testPersistenceSingleNodeDeleteRealm() {
...
+        try {
+            List<String> offlineSessionIds2 = createOfflineSessions(realmId2, userIds2);
+            assertOfflineSessionsExist(realmId2, offlineSessionIds2);
+
+            // Simulate server restart
+            reinitializeKeycloakSessionFactory();
+
+            assertOfflineSessionsExist(realmId2, offlineSessionIds2);
+
+            withRealm(realmId2, (session, realm) -> new RealmManager(session).removeRealm(realm));
```
Comment (mhajas) [inline testsuite/model/src/test/java/org/keycloak/testsuite/model/session/OfflineSessionPersistenceTest.java:149]: Shouldn't we use `session.reamls().removeRealm()` method here?

**Review by mhajas — COMMENTED** [review]
Thanks for the test, I added one question

File: testsuite/model/src/test/java/org/keycloak/testsuite/model/session/OfflineSessionPersistenceTest.java (hunk 2 above)
Comment (martin-kanis) [reply testsuite/model/src/test/java/org/keycloak/testsuite/model/session/OfflineSessionPersistenceTest.java]: Sure, changed.

**Review by mhajas — DISMISSED** [review]
Thanks for the PR. LGTM

**Review by hmlnarik — APPROVED** [review]
Thanks for the fix!

**Pull request merged**